calendar/store/user_store.go

```diff
@@ -295,6 +305,79 @@ func (s *pluginStore) StoreUserActiveEvents(mattermostUserID string, events []st
 	return kvstore.StoreJSON(s.userKV, mattermostUserID, u)
 }
 
+// RefreshAndStoreToken checks whether the current access token is expired or not. If it is,
+// then it refreshes the token and stores the new pair of access and refresh tokens in kv store.
+func (s *pluginStore) RefreshAndStoreToken(token *oauth2.Token, oconf *oauth2.Config, mattermostUserID string) (*oauth2.Token, error) {
```

Comment:
@Kshitij-Katiyar This is where much of the functionality was added in this PR. I'm thinking we can test this function and `CheckUserConnected` directly. What do you think?

We basically want to emulate this in a test since this is the code that is applied throughout the PR:

```go
	if !c.tokenHelpers.CheckUserConnected(c.mattermostUserID) {
		c.Logger.Warnf(LogUserInactive, c.mattermostUserID)
		return nil, errors.New(ErrorUserInactive)
	}
	err := someCallToAPI // this can be whatever we want in the test
	if err != nil {
		c.tokenHelpers.DisconnectUserFromStoreIfNecessary(err, c.mattermostUserID)
```

We can use a "real" KV store in the test with one of the following strategies:
- https://gist.github.com/mickmister/f2ac1c99c63291692b18fdae715338af
- https://github.com/mattermost/mattermost-plugin-github/pull/745

Reply:
@wiggin77 Writing test cases for these functions without refactoring is challenging because many of the functions used within them are standalone and not implemented by any interface, making it difficult to mock them.

One workaround is to use monkeyPatch. However, this approach isn't currently used in the plugin, and introducing it just for this test case might not be ideal. Additionally, no test cases have been added for any `store` or `msgraph` functions so far, possibly because adding them is not straightforward.

Reply:
This method can be unit tested with a memory store.

Reply:
Hi @wiggin77,
I looked into the memoryStore and its implementation in Mattermost. It seems useful for mocking the config store in Mattermost and other similar stores within Mattermost plugins. However, I’m unsure how it applies to our current use case. Could you provide some guidance on this?

Reply:
A solution needs to be found here.  If the implementation does not lend itself to writing clean unit tests, then the implementation needs to be refactored.